src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Map.cs

```diff
@@ -0,0 +1,28 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+using System.Buffers.Binary;
+
+namespace System.Security.Cryptography.Encoding.Tests.Cbor
+{
+    internal partial class CborReader
+    {
+        public ulong? ReadStartMap()
+        {
+            CborInitialByte header = PeekInitialByte(expectedType: CborMajorType.Map);
+            ulong arrayLength = checked((ulong)ReadUnsignedInteger(header, out int additionalBytes));
+            AdvanceBuffer(1 + additionalBytes);
+            _remainingDataItems--;
+
+            PushDataItem(CborMajorType.Map, 2 * arrayLength);
```

Comment:
Are there any simple tests we can do with the value for either maps or arrays to decide we're broken?

e..g since there's nothing that compresses two values into one byte, the encoded length (or, for maps, double-the-length) can't exceed the remaining bytes in the payload.

Consider a caller that allocates an array of length ReadStartArray. An obviously fraudulent large value can make them OOM (of course, a maximally legal value might also do that; but let's prevent them from being told "2 billion entries" when the payload is only 12 bytes long)

Reply:
Sounds reasonable, although I've already written a handful of tests that rely on this being treated as valid (that is it won't fail at the `ReadStartArray()` phase). Would it make sense to expose as an opt-out check?

Reply:
What's the scenario (aside from existing tests codifying the behavior) for allowing it to succeed here when the value passed to PushDataItem exceeds the remaining buffer length?  (That is, why have an option, instead of just failing?).

I don't think I'm advocating for a full Skip-Walk to verify the length (maybe _that_ would be an option someone wants), just avoiding egregious errors.

Reply:
> What's the scenario (aside from existing tests codifying the behavior) for allowing it to succeed here when the value passed to PushDataItem exceeds the remaining buffer length?

None really. It really shouldn't be exposed, but I can't think of a good way to balance this (we could switch it off using reflection, if that counts as a good way).

> I don't think I'm advocating for a full Skip-Walk to verify the length (maybe that would be an option someone wants), just avoiding egregious errors.

Yes, that's clear. One could argue that since this is a measure guarding against allocations that could be made by consumers of the Reader, they could easily make that check themselves, assuming they needed to be this proactive. OTOH we don't do this check in APIs where we allocate buffers, i.e. `ReadTextString()` and `ReadByteString()`, however that's the exactly the type of scenario that `TryReadTextString()` and `TryReadByteString()` are meant to mitigate.

Another thing to note is that adopting this check further moves this implementation from a stream-based API.